Podcast Player - make feed fetch call cancellable and cancel on Block removal.

getdave opened the pull request:
Currently the Podcast Player Block uses apiFetch to trigger a Ajax request for the feed data. Once it [the fetch promise] resolves then state and attributes updates are triggered.

If, whilst the network request is in progress, the Block is "unmounted" then the console will show an error along the lines of

> Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.

This is because the .then handler on the fetch still triggers even though the Block no longer exists. As the handler updates state and attributes the console triggers the above error regarding state updates on unmounted components.

To fix this we make the `apiFetch` call cancell-able (via a wrapper util) and then ensure we cancel and do not update state if the Block has been unmounted.

Fixes https://github.com/Automattic/jetpack/issues/15214

## Changes proposed in this Pull Request:
* Wrap `apiFetch` to make it possible to cancel it.
* Add ref to track cancellable promise instance.
* Add `useEffect` to `.cancel()` the fetch promise if the Block unmounts.
* Add conditional to the `.catch()` handler to check for `.isCancelled` and terminate handler execution early to avoid state updates.

## Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Fixes existing bug in Podcast Player Block

## Testing instructions:

Annoyingly there seems to be a bug in Gutenberg with the Popover component. Therefore when "removing" the BLock below you will need to click to place the cursor in the editor immediate following the Block and then hit the `BACKSPACE` key to remove the block. Removing using the dropdown many on the Block itself will trigger a warning which is **unrelated to this Block**.

#### On Feature Branch

* Insert Podcast Block.
* In Network Tools switch to "Slow 3G mode".
* Add a valid feed URL and submit.
* Immediately (before the fetch request resolves) delete the Block (see notes above).
* See error in console relating to state updates on unmounted component.

#### On This Branch

* Insert Podcast Block.
* In Network Tools switch to "Slow 3G mode".
* Add a valid feed URL and submit.
* Immediately (before the fetch request resolves) delete the Block (see notes above).
* See no errors.

If you enable `debug` mode you will also see a notice informing you that:

> Block was unmounted during fetch

 

#### Proposed changelog entry for your changes:
* Fix React error regarding state updates on unmounted component if Block is removed during the fetching of the Podcast feed.

getdave committed:
Make fetch call cancellable and cancel on unmount.

retrofox commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -81,21 +82,30 @@ const PodcastPlayerEdit = ( {
+			);
+
+			cancellableFetch.current.promise.then(
 				data => {
```
`isCancel` seems to be present in the `data` object too. Does it make sense to check if the request has been canceled here in the same way when an error happens?

getdave replied:
Great spot. You are absolutely correct. I've fixed this also.

retrofox commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -106,6 +116,14 @@ const PodcastPlayerEdit = ( {
 		[ createErrorNotice ]
 	);
 
+	useEffect( () => {
+		return () => {
+			if ( cancellableFetch && cancellableFetch.current && cancellableFetch.current.cancel ) {
+				cancellableFetch.current.cancel();
+			}
+		};
+	}, [] );
+
```
Simple suggestion to make it clearer. Feel free to ignore it :-)

```suggestion
	useEffect( () => (
		() => get( cancellableFetch, [ 'current', 'cancel' ] )
			? cancellableFetch.current.cancel()
			: noop()
	), [] );
```

Don't forget to add the dependencies if you agree with these changes.

getdave replied:
OK I'm not 100% sure on this but I've ended up using Optional Chaining as an alternative here.

Supported as of https://github.com/Automattic/wp-calypso/blob/c3608b5c2b656cad1209f52b0e2229838c2592ea/packages/calypso-build/CHANGELOG.md#510

retrofox reviewed (approved):
I left only one comment about checking the `isCancel` state. Maybe it doesn't make sense, though. 

Tested and it works as expected.

![image](https://user-images.githubusercontent.com/77539/78139309-abd56d80-73fe-11ea-8ad8-725a54739394.png)

Nice improvement, Dave. 💯

We can wait for more feedback from other folks.

getdave committed:
Use optional chaining to reduce complexity

Seems to have been added to Calypso build here

https://github.com/Automattic/wp-calypso/blob/c3608b5c2b656cad1209f52b0e2229838c2592ea/packages/calypso-build/CHANGELOG.md#510

…so should be safe to used as it will be polyfilled.

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

getdave committed:
Fix to also exit early on happy .then() path if promise has been cancelled

Kudos to https://github.com/Automattic/jetpack/pull/15228#discussion_r401579732

getdave commented:
As this is merging into the feature branch we won't wait on Jetpack Crew.
